src/excited_workflow/train_fluxnet_models.py

```diff
@@ -0,0 +1,339 @@
+"""Workflow to produce the Fluxnet-based ML model(s)."""
+import shutil
+from collections.abc import Callable
+from collections.abc import Iterable
+from dataclasses import dataclass
+from datetime import datetime
+from pathlib import Path
+from typing import Any
+
+import numpy as np
+import onnxmltools
+import pycaret.regression
+import xarray as xr
+from skl2onnx import convert_sklearn
+from skl2onnx.common.data_types import DoubleTensorType
+from sklearn.model_selection import BaseCrossValidator
+
+from excited_workflow import fluxnet_site_extraction
+from excited_workflow.common_utils import write_model_variables
+from excited_workflow.config import PATHS_CFG
+from excited_workflow.source_datasets import datasets
+
+
+def calculate_era5_derived_vars(era5_dataset: xr.Dataset) -> xr.Dataset:
+    """Calculate variables, based on the ERA5 dataset.
+
+    Variables calculated:
+        ssr_6hr: six hour rolling mean of the incoming shortwave solar radiation.
+        day_of_year: The day of year.
+        hour: The hour of day.
+        t2m_1w_rolling: 1-week rolling mean of the air temperature.
+        dewpoint_depression_1w_rolling: 1-week rolling mean of the dewpoint depression.
+        mean_temp: Mean temperature over time (over the entire time period available).
+        mean_dewpoint_depression: Mean dewpoint depression.
+    """
+    era5_dataset = era5_dataset.copy(deep=True)
+    era5_dataset["ssr_6hr"] = (
+        era5_dataset["ssr"].rolling(time=6, center=True).mean()
+    )
+
+    era5_dataset["day_of_year"] = era5_dataset["time"].dt.dayofyear
+    era5_dataset["hour"] = era5_dataset["time"].dt.hour
+
+    era5_dataset["t2m_1w_rolling"] = (
+        era5_dataset["t2m"].rolling(time=7 * 24, center=True).mean()
+    )
+    era5_dataset["mean_air_temperature"] = era5_dataset["t2m"].mean(dim="time")
+
+    era5_dataset["mean_dewpoint_depression"] = era5_dataset["t2m"].mean(
+        dim="time"
+    ) - era5_dataset["d2m"].mean(dim="time")
+
+    era5_dataset["dewpoint_depression_1w_rolling"] = (
+        era5_dataset["t2m"].rolling(time=7 * 24, center=True).mean()
+        - era5_dataset["d2m"].rolling(time=7 * 24, center=True).mean()
+    )
+    return era5_dataset
+
+
+def compute_respiration(fluxnet_data: xr.Dataset) -> xr.Dataset:
+    """Compute respiration from fluxnet GPP and NEE."""
+    fluxnet_data = fluxnet_data.copy(deep=True)
+    fluxnet_data["resp"] = fluxnet_data["GPP_NT_VUT_REF"] + fluxnet_data["NEE_VUT_REF"]
+    fluxnet_data["resp"].attrs = {
+        "units": "umolCO2 m-2 s-1",
+        "long_name": "ecosystem_respiration",
+    }
+    fluxnet_data["GPP_NT_VUT_REF"].attrs = {
+        "units": "umolCO2 m-2 s-1",
+        "long_name": "gross_primary_production",
+    }
+    fluxnet_data["NEE_VUT_REF"].attrs = {
+        "units": "umolCO2 m-2 s-1",
+        "long_name": "net_ecosystem_exchange",
+    }
+    return fluxnet_data
+
+
+def extract_sites_from_datasets(
```

Comment:
The difference is mainly due to one being half-hourly (ERA5) and the other monthly datasets. The half-hourly case also needs to be performed in a pre-processing step as it would otherwise take way too long.

However, it is probably better to reorganize the code so the two extraction functions are in the same module.

Reply:
Moved it to the module `fluxnet_site_extraction.py`